diff --git a/src/http/ngx_http_upstream.c b/src/http/ngx_http_upstream.c
index f5db65338..5905e0842 100644
--- a/src/http/ngx_http_upstream.c
+++ b/src/http/ngx_http_upstream.c
@@ -8,6 +8,9 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_http.h>
+#if (NGX_HTTP_APISIX)
+#include <ngx_http_apisix_module.h>
+#endif
 
 
 #if (NGX_HTTP_CACHE)
@@ -2977,6 +2980,12 @@ ngx_http_upstream_process_trailers(ngx_http_request_t *r,
         return NGX_OK;
     }
 
+#if (NGX_HTTP_APISIX)
+    if (!ngx_http_apisix_is_upstream_pass_trailers(r)) {
+        return NGX_OK;
+    }
+#endif
+
     part = &u->headers_in.trailers.part;
     h = part->elts;
 
