diff --git src/stream/ngx_stream_proxy_module.c src/stream/ngx_stream_proxy_module.c
index 82dca1e..7cc9fa4 100644
--- src/stream/ngx_stream_proxy_module.c
+++ src/stream/ngx_stream_proxy_module.c
@@ -8,6 +8,9 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_stream.h>
+#if (NGX_STREAM_APISIX)
+#include <ngx_stream_apisix_module.h>
+#endif
 
 
 typedef struct {
@@ -2223,7 +2226,11 @@ ngx_stream_proxy_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
     ngx_conf_merge_ptr_value(conf->ssl_conf_commands,
                               prev->ssl_conf_commands, NULL);
 
+#if (NGX_STREAM_APISIX)
+    if (ngx_stream_proxy_set_ssl(cf, conf) != NGX_OK) {
+#else
     if (conf->ssl_enable && ngx_stream_proxy_set_ssl(cf, conf) != NGX_OK) {
+#endif
         return NGX_CONF_ERROR;
     }
 
