diff --git src/ngx_http_lua_ssl_certby.c src/ngx_http_lua_ssl_certby.c
index 0901f06e..a5714226 100644
--- src/ngx_http_lua_ssl_certby.c
+++ src/ngx_http_lua_ssl_certby.c
@@ -1467,9 +1467,16 @@ ngx_http_lua_ssl_verify_callback(int ok, X509_STORE_CTX *x509_store)
 #endif
 
 
+static int
+ngx_http_lua_ssl_verify_reject_in_handshake_callback(int ok, X509_STORE_CTX *x509_store)
+{
+    return ok;
+}
+
+
 int
 ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *client_certs,
-    void *trusted_certs, int depth, char **err)
+    void *trusted_certs, int depth, int reject_in_handshake, char **err)
 {
 #ifdef LIBRESSL_VERSION_NUMBER
 
@@ -1517,7 +1524,14 @@ ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *client_certs,
 
     /* enable verify */
 
-    SSL_set_verify(ssl_conn, SSL_VERIFY_PEER, ngx_http_lua_ssl_verify_callback);
+    if (reject_in_handshake) {
+        SSL_set_verify(ssl_conn,
+                       SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT ,
+                       ngx_http_lua_ssl_verify_reject_in_handshake_callback);
+
+    } else {
+        SSL_set_verify(ssl_conn, SSL_VERIFY_PEER, ngx_http_lua_ssl_verify_callback);
+    }
 
     /* set depth */
 
