diff --git src/http/modules/ngx_http_proxy_module.c src/http/modules/ngx_http_proxy_module.c
index 3f05235..41c2313 100644
--- src/http/modules/ngx_http_proxy_module.c
+++ src/http/modules/ngx_http_proxy_module.c
@@ -8,6 +8,9 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_http.h>
+#if (NGX_HTTP_APISIX)
+#include "ngx_http_apisix_module.h"
+#endif
 
 
 #define  NGX_HTTP_PROXY_COOKIE_SECURE           0x0001
@@ -1013,7 +1016,11 @@ ngx_http_proxy_handler(ngx_http_request_t *r)
 
     u->accel = 1;
 
+#if (NGX_HTTP_APISIX)
+    if (!ngx_http_apisix_is_request_buffering(r, plcf->upstream.request_buffering)
+#else
     if (!plcf->upstream.request_buffering
+#endif
         && plcf->body_values == NULL && plcf->upstream.pass_request_body
         && (!r->headers_in.chunked
             || plcf->http_version == NGX_HTTP_VERSION_11))
