diff --git src/core/ngx_thread_pool.c src/core/ngx_thread_pool.c
index 7fb0f7f..95081a6 100644
--- src/core/ngx_thread_pool.c
+++ src/core/ngx_thread_pool.c
@@ -587,6 +587,9 @@ ngx_thread_pool_init_worker(ngx_cycle_t *cycle)
     ngx_thread_pool_conf_t   *tcf;
 
     if (ngx_process != NGX_PROCESS_WORKER
+#if HAVE_PRIVILEGED_PROCESS_PATCH
+        && !ngx_is_privileged_agent
+#endif
         && ngx_process != NGX_PROCESS_SINGLE)
     {
         return NGX_OK;
