diff --git lib/ngx/pipe.lua lib/ngx/pipe.lua
index 3757ab7..413a10e 100644
--- lib/ngx/pipe.lua
+++ lib/ngx/pipe.lua
@@ -47,8 +47,7 @@ typedef struct {
     ngx_http_lua_pipe_t    *pipe;
 } ngx_http_lua_ffi_pipe_proc_t;
 
-int ngx_http_lua_ffi_pipe_spawn(ngx_http_request_t *r,
-    ngx_http_lua_ffi_pipe_proc_t *proc,
+int ngx_http_lua_ffi_pipe_spawn(ngx_http_lua_ffi_pipe_proc_t *proc,
     const char *file, const char **argv, int merge_stderr, size_t buffer_size,
     const char **environ, u_char *errbuf, size_t *errbuf_size);
 
@@ -625,9 +624,8 @@ do
 
         local errbuf = get_string_buf(ERR_BUF_SIZE)
         local errbuf_size = get_size_ptr()
-        local r = get_request()
         errbuf_size[0] = ERR_BUF_SIZE
-        local rc = C.ngx_http_lua_ffi_pipe_spawn(r, proc, exe, proc_args,
+        local rc = C.ngx_http_lua_ffi_pipe_spawn(proc, exe, proc_args,
                                                  merge_stderr, buffer_size,
                                                  proc_envs, errbuf, errbuf_size)
         if rc == FFI_ERROR then
