diff --git src/http/ngx_http_request.c src/http/ngx_http_request.c
index 013b715..a729693 100644
--- src/http/ngx_http_request.c
+++ src/http/ngx_http_request.c
@@ -754,6 +754,11 @@ ngx_http_ssl_handshake(ngx_event_t *rev)
                 return;
             }
 
+#if (TONGSUO_VERSION_NUMBER && NGX_HTTP_APISIX)
+            // FIXME: add option later
+            SSL_enable_ntls(c->ssl->connection);
+#endif
+
             ngx_reusable_connection(c, 0);
 
             rc = ngx_ssl_handshake(c);
