diff --git src/ngx_http_lua_ssl_certby.c src/ngx_http_lua_ssl_certby.c
index 6ed2f3f..c46cc91 100644
--- src/ngx_http_lua_ssl_certby.c
+++ src/ngx_http_lua_ssl_certby.c
@@ -1346,9 +1346,16 @@ ngx_http_lua_ssl_verify_callback(int ok, X509_STORE_CTX *x509_store)
 }
 
 
+static int
+ngx_http_lua_ssl_verify_reject_in_handshake_callback(int ok, X509_STORE_CTX *x509_store)
+{
+    return ok;
+}
+
+
 int
 ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *ca_certs,
-    int depth, char **err)
+    int depth, int reject_in_handshake, char **err)
 {
     ngx_http_lua_ctx_t          *ctx;
     ngx_ssl_conn_t              *ssl_conn;
@@ -1388,7 +1395,14 @@ ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *ca_certs,
 
     /* enable verify */
 
-    SSL_set_verify(ssl_conn, SSL_VERIFY_PEER, ngx_http_lua_ssl_verify_callback);
+    if (reject_in_handshake) {
+        SSL_set_verify(ssl_conn,
+                       SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT ,
+                       ngx_http_lua_ssl_verify_reject_in_handshake_callback);
+
+    } else {
+        SSL_set_verify(ssl_conn, SSL_VERIFY_PEER, ngx_http_lua_ssl_verify_callback);
+    }
 
     /* set depth */
 
